content/blog/2019-11-13-llvm-profiling/index.md

```diff
@@ -0,0 +1,336 @@
+
+
+
+
+
+
+## Hardest Parts
+
+1. For people who are new to LLVM, instructions are hard to find and follow. Searching on google can help if you know what you're looking for. It's difficult to get helpful information unless our search is confined to existing functions. Even though LLVM documentation is pretty exhaustive, it has too many functions to go through and the lack of examples can be off putting. Tutorials and existing backbone codes on Github can be really handy in these scenarios, which we took advantage of. It not only helped us implement specific functions like `printf` but also establish a structure to our IR pass. 
+2. String manipulations: I am not sure if this is the right term to use, but LLVM seems to have 2 string types - Twine and StringRef. `getName` on a function returns a StringRef. In order to make a custom name I perform `F.getName()+"name"` which returns a Twine. But the function `getGlobalVariable` only accepts StringRef. Twine has a function `str` which can be used to convert it into string. Even though this is a straight forward solution, it ended up taking time to figure out the problem and looking into the documentation of these classes.
+3. Setting up and running benchmarks: "It's all fun and games until you run your IR pass on real programs" - anonymous. We faced issues setting up benchmarks like PARSEC on Mac. Embench had multiple source files to compile which ran into trouble partly due to our IR pass not being thoughtfully written. We were defining global variables in all the files irrespective of it being a function/utility file or the main source file. We ended up using [Phoenix](https://github.com/kozyraki/phoenix) which worked well on Linux but was not meant for Mac. Hence for doing our LLVM pass, we had to install and update libraries on Linux machines.
+
+
+
+## Evaluation and results
+
+### Gcov profiling tool
+
+To validate and benchmark our profiling results, we use `gcov` testing tool which can be used as a profiler to give performance statistics. We first compile the code using gcc flags required for `gcov` : `gcc -fprofile-arcs -ftest-coverage foo.c` . Now we run `gcov` with relevant flags to give us statistics to compare with our profiler:
+```
+gcov -b -c -f foo.c
+```
+We ran the [Phoenix](https://github.com/kozyraki/phoenix) benchmark suite and used `gcov` to profile statistics of function calls. The makefile initially had optimization flag -O3 but this might lead to incorrect numbers of function calls due to optimizations, hence we compile without any flags.
+
+#### Kmeans
```

Comment:
Any particular reason you picked Kmeans as the one benchmark to study in more detail, or did you just pick one arbitrarily? (It might be worth clarifying this in the text.)

Reply:
Picked arbitrarily. Mentioned this in a line. Thanks!